overlord/snapshotstate/backend/trash.go

```diff
@@ -0,0 +1,81 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package backend
+
+import (
+	"encoding/json"
+	"os"
+
+	"github.com/snapcore/snapd/logger"
+)
+
+// Trash stores information that can be used to cleanly revert (or
+// finish cleaning up) a snapshot Restore.
+//
+// This is useful when a Restore is part of a chain of operations, and
+// a later one failing necessitates undoing the Restore.
+type Trash struct {
+	Done    bool     `json:"done,omitempty"`
+	Created []string `json:"created,omitempty"`
+	Moved   []string `json:"moved,omitempty"`
+	// Config is here for convenience; this package doesn't touch it
+	Config *json.RawMessage `json:"config,omitempty"`
+}
+
+// check that you're not trying to use trash info twice; data loss imminent.
```

Comment:
The documentation is phrased unconventionally and also a bit confusing in the sense that it's okay to _use_ the information twice, otherwise we'd not have json fields above. There's something else that cannot be done twice, but it's not mentioned here.

Reply:
dropping the function made this point moot \o/